File: 15-3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> ans;
        sort(nums.begin(), nums.end());
        for(int i = 0; i < n; i++) {
            if(i != 0 && nums[i] == nums[i - 1]) continue;
            // int num1 = nums[i];
            int j = i + 1;
            int k = n - 1;
            while(j < k) {
                int sum = nums[i] + nums[j] + nums[k]; 
                if(sum == 0) {
                    vector<int> v = {nums[i], nums[j], nums[k]};
                    ans.push_back(v);
                    j++;
                    k--;
                    while(j < n && nums[j] == nums[j - 1]) j++;
                    while(k > j && nums[k] == nums[k + 1]) k--;
                } 
                else if(sum > 0) k--;
                else j++;
            }
        }
        return ans;
    }
};

File: 15-3sum/3sum.cs
public class Solution {
    public IList<IList<int>> ThreeSum(int[] nums) {
        IList<IList<int>> li = new List<IList<int>>();
        HashSet<(int, int, int)> hs = new HashSet<(int, int, int)>();
        int n = nums.Length;
        Array.Sort(nums);
        for(int i = 0; i < n; i++) {
            int newTarget = -nums[i];
            Dictionary<int, int> dt = new Dictionary<int, int>();
            for(int j = i + 1; j < n; j++) {
                if(dt.ContainsKey(newTarget - nums[j])) {
                    List<int> v = new List<int>();
                    hs.Add((nums[i], newTarget - nums[j], nums[j]));
                }
                dt[nums[j]] = j;
            }
        }
        HashSet<string> hst = new HashSet<string>();
        
        foreach((int, int, int) x in hs) {
            List<int> v = new List<int>();
            v.Add(x.Item1);
            v.Add(x.Item2);
            v.Add(x.Item3);
            li.Add(v);
        }
        return li;
    }
}